In [5]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [6]:
df=pd.read_csv(r'C:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\data\processed\hebin_p1w2_features.csv')

In [7]:
feature_cols = ["age","pregnancies","glucose","blood_pressure",
                "skin_thickness","insulin","bmi","diabetes_pedigree",
                "bmi_category","glucose_band"]
x = df[feature_cols].copy()
y = df["outcome"]
x

,age,pregnancies,glucose,blood_pressure,skin_thickness,insulin,bmi,diabetes_pedigree,bmi_category,glucose_band
0,79,0,124.000000,71.0,30.0,124.0,23.4,0.588,normal,prediabetic
1,37,0,153.000000,85.0,24.0,42.0,33.9,0.192,obese,diabetic
2,39,0,142.000000,68.0,22.0,159.0,26.9,0.777,overweight,diabetic
3,68,7,121.000000,88.0,26.0,124.0,29.0,1.217,overweight,prediabetic
4,75,0,107.000000,84.0,31.0,101.0,21.2,0.278,normal,prediabetic
...,...,...,...,...,...,...,...,...,...,...
945,64,2,117.855165,81.0,34.0,124.0,40.3,0.641,obese,prediabetic
946,71,0,140.000000,64.0,37.0,124.0,33.4,0.116,obese,diabetic
947,40,0,130.000000,73.0,26.0,258.0,20.1,0.709,normal,diabetic
948,52,4,123.000000,76.0,22.0,189.0,38.9,0.540,obese,prediabetic


In [8]:
x=pd.get_dummies(x,drop_first=True)
x.shape

(950, 13)

In [9]:
x_train,x_test,y_train,y_test= train_test_split(
    x,y,test_size=0.2,random_state=42,stratify=y)

In [10]:
scaler = StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)#Calculates the mean and standard deviation from x train
x_test_scaled=scaler.transform(x_test)#Uses those values to standardize x_train.

In [11]:
logreg = LogisticRegression(max_iter=1000).fit(x_train_scaled, y_train)
dtree = DecisionTreeClassifier(max_depth=3,random_state=42).fit(x_train_scaled, y_train)
knn = KNeighborsClassifier(n_neighbors=5).fit(x_train_scaled, y_train)



In [12]:
print('LogisticRegression',accuracy_score(y_test,logreg.predict(x_test_scaled)))
print('DecisionTreeClassifier',accuracy_score(y_test,dtree.predict(x_test)))
print('kneighborsclassifier',accuracy_score(y_test,knn.predict(x_test_scaled)))

LogisticRegression 0.7526315789473684
DecisionTreeClassifier 0.2894736842105263
kneighborsclassifier 0.6842105263157895


c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


In [16]:
models = [
    ("Logistic Regression", logreg, x_test_scaled),
    ("Decision Tree (d=3)", dtree, x_test),
    ("KNN (k=25)", knn, x_test_scaled)
]

for name, m, Xt in models:
    print(f"{name:22s} accuracy = {accuracy_score(y_test, m.predict(Xt)):.4f}")

print("\nTest set:", len(y_test), "patients |",
      int(np.sum(y_test)), "of them diabetic")

Logistic Regression    accuracy = 0.7526
Decision Tree (d=3)    accuracy = 0.2895
KNN (k=25)             accuracy = 0.6842

Test set: 190 patients | 55 of them diabetic


c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


In [ ]:
def cm_row(name,model,xt):
   cm=confusion_matrix(y_test,model.predict(xt))
   tn,fp,fn,tp=cm.ravel()
   return{"model":name,"correct_negatives":int(tn),"false_alarams":int(fp),
          "missed_patients":int(fn),"found_patients":int(tp),
          "accuracy": round(model.score(xt, y_test), 4)
    } 
pd.DataFrame([
    cm_row("Logistic Regression",logreg, x_test_scaled),
    cm_row("Decision Tree (d=3)", dtree, x_test),
    cm_row("KNN (k=25)", knn, x_test_scaled),
])

c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(
c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


,model,correct_negatives,false_alarams,missed_patients,found_patients,accuracy
0,Logistic Regression,127,8,39,16,0.7526
1,Decision Tree (d=3),0,135,0,55,0.2895
2,KNN (k=25),113,22,38,17,0.6842


Build a confusion matrices for every model

In [17]:
rows = []
for name,m,Xt in models:
    cm = confusion_matrix(y_test,m.predict(Xt))
    tn, fp, fn, tp = cm.ravel()
    rows.append({"models": name, "correct_negative": int(tn),"False_alarms":int(fp),
                 "missed_patients":int(fn),"found_patients":int(tp),
                 "accuracy":accuracy_score(y_test,m.predict(Xt))})
    metrices = pd.DataFrame(rows)
metrices

c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(
c:\Users\user\OneDrive\Desktop\5thsem_ojt\project1_diabetes\my_env\Lib\site-packages\sklearn\utils\validation.py:2820: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


,models,correct_negative,False_alarms,missed_patients,found_patients,accuracy
0,Logistic Regression,127,8,39,16,0.752632
1,Decision Tree (d=3),0,135,0,55,0.289474
2,KNN (k=25),113,22,38,17,0.684211


Correct Negative (TN) = 127 → 127 patients were correctly predicted as negative.
False Alarm (FP) = 8 → 8 patients were predicted positive, but were actually negative.
Missed Patients (FN) = 39 → 39 patients actually had the disease, but the model predicted negative.
Found Patients (TP) = 16 → 16 patients with the disease were correctly identified.
Accuracy = 75.26%

Logistic Regression is dominated by missed patients, Decision Tree is dominated by false alarms, and KNN is dominated by missed patients

Score the model that never says yes

In [24]:
from sklearn.dummy import DummyClassifier

dummy=DummyClassifier(strategy='most_frequent')
dummy.fit(x_train,y_train)
y_pred=dummy.predict(x_test)
confusion_matrix(y_test,y_pred)

array([[135,   0],
       [ 55,   0]])

In [25]:
print("Dummy Classifier Accuracy:", accuracy_score(y_test, y_pred))

Dummy Classifier Accuracy: 0.7105263157894737


In [28]:
y_test.value_counts(normalize=True)

outcome
0    0.710526
1    0.289474
Name: proportion, dtype: float64